bibtex-completion.el

```diff
@@ -1537,41 +1537,64 @@ BibTeX files.  If this fails, return nil."
            (require 'reftex-cite nil t)
            (ignore-errors (reftex-get-bibfile-list)))))
 
+(defun bibtex-completion-get-key-bibtex ()
+  "Return the key of the BibTeX entry at point, nil otherwise.
+This function can be used by `bibtex-completion-key-at-point' to
+find the key of the BibTeX entry at point in a BibTeX-mode
+buffer."
+  (when (eq major-mode 'bibtex-mode)
+    (save-excursion
+      (bibtex-beginning-of-entry)
+      (and (looking-at bibtex-entry-maybe-empty-head)
+           (bibtex-key-in-head)))))
+
+(defun bibtex-completion-get-key-reftex ()
+  "Return the key of the BibTeX entry at point, nil otherwise.
+This function can be used by `bibtex-completion-key-at-point' to
+find the key of the BibTeX entry at point in a LaTeX buffer."
+  (when (require 'reftex-parse nil t)
+    (save-excursion
+      (skip-chars-backward "[:space:],;}")
+      (let ((macro (reftex-what-macro 1)))
+        (and (stringp (car macro))
+             (string-match "\\`\\\\cite\\|cite\\'" (car macro))
+             ;; allow '_' in citekeys
+             (let ((temp-syn-table (make-syntax-table)))
+               (modify-syntax-entry ?_ "_" temp-syn-table)
+               (with-syntax-table temp-syn-table
+                 (thing-at-point 'symbol))))))))
+
+(defun bibtex-completion-get-key-org ()
+  "Return the key of the BibTeX entry at point, nil otherwise.
+This function can be used by `bibtex-completion-key-at-point' to
+find the key of the BibTeX entry at point in an Org-mode buffer."
+  (when (eq major-mode 'org-mode)
+    (let (key)
+      (and (setq key (org-entry-get nil
+                                    (if (boundp 'org-bibtex-key-property)
+                                        org-bibtex-key-property
+                                      "CUSTOM_ID")
+                                    t))
+           ;; KEY may be the empty string the the property is
+           ;; present but has no value
+           (> (length key) 0)
+           key))))
+
+(defvar bibtex-completion-key-at-point-functions
```

Comment:
Went for a `defvar` instead of a `defcustom` because it's geared towards packages rather than users.